### PR TITLE
Implement support for pointxyz and range leaves

### PR DESCRIPTION
**Pull request by knzivid**

### CODE REVIEW

File: zivid_camera/test/test_zivid_camera.cpp [hunk 1]
```diff
@@ -460,6 +465,67 @@ TEST_F(CaptureOutputTest, testCapturePointsXYZ)
...
+  const auto& point_cloud = points_sub.lastMessage();
+  ASSERT_TRUE(point_cloud.has_value());
+
+  const auto point_cloud_sdk = captureViaSDK(Zivid::Settings{
+      Zivid::Settings::Acquisitions{ Zivid::Settings::Acquisition{} },
+      Zivid::Settings::RegionOfInterest::Box{
+          Zivid::Settings::RegionOfInterest::Box::Enabled::yes,
+          Zivid::Settings::RegionOfInterest::Box::PointO{ 0, 0, 600 },
+          Zivid::Settings::RegionOfInterest::Box::PointA{ 10, 0, 600 },
+          Zivid::Settings::RegionOfInterest::Box::PointB{ 0, 10, 600 },
```
[the remark this answers is not in the document]
Comment (apartridge) [reply zivid_camera/test/test_zivid_camera.cpp:499]: Also, please verify that the ROI here is defined such that it actually filters point. If it fex covered the entire scene then we would not really test anything. If you need a sameple ROI you can search for regionOfInterestBoxDivideHorizontal in SDK.

[the remark this answers is not in the document]
Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp:499]: 1. Offset all X coordinates by 1 such that all components of a coordinate are unique
2. Without ROI, the number of NaNs reduce to 118k. With ROI, it becomes 2.3M. It is filtering away points.

**Comment by knzivid** [comment]
RFNR

File: zivid_camera/test/test_zivid_camera.cpp [hunk 2]
```diff
@@ -460,6 +465,69 @@ TEST_F(CaptureOutputTest, testCapturePointsXYZ)
   }
 }
 
+#if ZIVID_CORE_VERSION_MAJOR >= 2 && ZIVID_CORE_VERSION_MINOR >= 9
```
Comment (apartridge) [inline zivid_camera/test/test_zivid_camera.cpp:468]: This tests looks incorrect, as it would break if we go to 3.0.

Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp:468]: can't we fix it after 3.0? I am not sure if we committed to 3.0 including all API breaking changes or just 2.10 for the next release

Comment (apartridge) [reply zivid_camera/test/test_zivid_camera.cpp:468]: I think we must fix this so it works as expected in the future. It could be that a 3.10 does not require much changes in ROS driver (if any). It should be quick to fix.

Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp:468]: It could be that we switch to a monotonic version number like https://github.com/llvm/llvm-project/blob/466d554dcab39c3d42fe0c5b588b795e0e4b9d0d/clang/lib/Headers/stdatomic.h#L56 for API 3.0 and not use split major-minor versions for this check?

Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp]: anyway, I added one more check for 3.0

**Review by apartridge — COMMENTED** [review]
PR looks good! I will just take a closer look at it tomorrow and test it a bit.

File: zivid_camera/test/test_zivid_camera.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (apartridge) [reply zivid_camera/test/test_zivid_camera.cpp]: I think you should also use different Z values, fex 590, 600, 610. The test is better that way because currently the test could hide a potential future bug where fex b_z and o_z got mixed up.

**Comment by apartridge** [comment]
The code is looking good, I tested it and it works as expected. I just added a few comments on the test. If you can fix those I will approve.

I think we can do the bump to 2.9 in a separate PR, as we also want to update README etc.

File: zivid_camera/src/settings_generator.cpp [hunk 3]
```diff
@@ -197,23 +198,35 @@ class DynamicReconfigureCfgGenerator
...
+    else
+    {
+      return ValueType{ 0 };
+    }
   }
 
   template <typename ValueType>
   auto convertValueToRosValue(ValueType value)
   {
     // Convert from our own setting value types to types that ROS params supports (double, int, bool)
```
[the remark this answers is not in the document]
Comment (knzivid) [reply zivid_camera/src/settings_generator.cpp]: removed

**Comment by knzivid** [comment]
Unless you have any other feedback, can you approve the PR?

**Comment by apartridge** [comment]
https://github.com/zivid/zivid-ros/pull/81#discussion_r1157007718

**Comment by knzivid** [comment]
Remember to unresolve the thread if you comment on them :)

File: zivid_camera/test/test_zivid_camera.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp]: Is this a blocker or a nice to have?

[the remark this answers is not in the document]
Comment (apartridge) [reply zivid_camera/test/test_zivid_camera.cpp]: This should be fixed, it should be easy to fix. We should ensure the tests are robust enough to cover potential corner case bugs in the future.

[the remark this answers is not in the document]
Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp]: Your call :)

[the remark this answers is not in the document]
Comment (knzivid) [reply zivid_camera/test/test_zivid_camera.cpp]: It does not make a difference to change Z a tiny bit because the extents are pretty large. We need a much smaller box that's more sensitive to Z.

**Comment by knzivid** [comment]
RFNR

**Comment by knzivid** [comment]
> PS: The CI looks a bit dead, probably because you are making this PR from your own fork?

Can you allow CI for me as an external contributor then?

**Review by apartridge — DISMISSED** [review]

**Review by apartridge — APPROVED** [review]

**Review by apartridge — APPROVED** [review]
